include/mbgl/gfx/polyline_generator.hpp

```diff
@@ -7,97 +7,823 @@
+                      const GeometryCoordinate& currentCoordinate,
+                      double& distance,
+                      const Point<double>& normal,
+                      double endLeft,
+                      double endRight,
+                      bool round,
+                      std::size_t startVertex,
+                      std::vector<TriangleElement>& triangleStore,
+                      std::optional<Distances> lineDistances,
+                      std::ptrdiff_t& e1,
+                      std::ptrdiff_t& e2,
+                      std::ptrdiff_t& e3);
 
-    void generate(const GeometryCoordinates& coordinates, const PolylineGeneratorOptions& options);
+template <class Vertices, class LayoutVertexFunc, class TriangleElement, class Distances>
+void addPieSliceVertex([[maybe_unused]] const std::size_t index,
+                       Vertices& vertices,
+                       LayoutVertexFunc layoutVertex,
+                       const GeometryCoordinate& currentVertex,
+                       double distance,
+                       const Point<double>& extrude,
+                       bool lineTurnsLeft,
+                       std::size_t startVertex,
+                       std::vector<TriangleElement>& triangleStore,
+                       std::optional<Distances> lineDistances,
+                       std::ptrdiff_t& e1,
+                       std::ptrdiff_t& e2,
+                       std::ptrdiff_t& e3);
+} // namespace detail
 
-private:
-    struct TriangleElement;
-
-    void addCurrentVertex(const GeometryCoordinate& currentCoordinate,
-                          double& distance,
-                          const Point<double>& normal,
-                          double endLeft,
-                          double endRight,
-                          bool round,
-                          std::size_t startVertex,
-                          std::vector<TriangleElement>& triangleStore,
-                          std::optional<PolylineGeneratorDistances> lineDistances);
-    void addPieSliceVertex(const GeometryCoordinate& currentVertex,
-                           double distance,
-                           const Point<double>& extrude,
-                           bool lineTurnsLeft,
-                           std::size_t startVertex,
-                           std::vector<TriangleElement>& triangleStore,
-                           std::optional<PolylineGeneratorDistances> lineDistances);
+} // namespace PolylineGenerator
 
-private:
-    Vertices& vertices;
-    LayoutVertexFunc layoutVertex;
-    Segments& segments;
-    CreateSegmentFunc createSegment;
-    GetSegmentFunc getSegment;
-    Indexes& indexes;
+template <class Vertices,
+          class LayoutVertexFunc,
+          class Segments,
+          class CreateSegmentFunc,
+          class GetSegmentFunc,
+          class Indexes>
+void PolylineGenerator::generate(Vertices& vertices,
+                                 LayoutVertexFunc layoutVertex,
```

Comment:
I think in this case (and for `addCurrentVertex`), it's probably preferable to stick with `std::function`. What I expect to happen here is each unique lambda instantiates a new instance of this template. Since there is a lot of code in this template function, each new lambda given duplicates all this code. That quickly sends the trade-off in the other direction towards `std::function` the moment more than one lambda is passed to this template. Since each lambda is unique and the template approach is an almost certain inline, that prevents the compiler from de-duping the 95+% of the function body that is otherwise identical.